apps/rush-lib/src/logic/buildCache/AzureStorageBuildCacheProvider.ts

```diff
@@ -0,0 +1,378 @@
+
+  private async _getSasQueryParametersAsync(terminal: Terminal): Promise<SASQueryParameters> {
+    let authorityHost: string;
+    switch (this._azureEnvironment) {
+      case AzureEnvironment.AzureCloud: {
+        authorityHost = 'https://login.microsoftonline.com';
+        break;
+      }
+
+      case AzureEnvironment.AzureChinaCloud: {
+        authorityHost = 'https://login.chinacloudapi.cn';
+        break;
+      }
+
+      case AzureEnvironment.AzureGermanCloud: {
+        authorityHost = 'https://login.microsoftonline.de';
+        break;
+      }
+
+      case AzureEnvironment.AzureUSGovernment: {
+        authorityHost = 'https://login.microsoftonline.us';
+        break;
+      }
+
+      default: {
+        throw new Error(`Unexpected Azure environment: ${this._azureEnvironment}`);
+      }
+    }
+
+    const deviceCodeCredential: DeviceCodeCredential = new DeviceCodeCredential(
+      undefined,
+      undefined,
+      (deviceCodeInfo: DeviceCodeInfo) => {
+        this._printMessageInBox(deviceCodeInfo.message, terminal);
+      },
+      { authorityHost: authorityHost }
+    );
+    const blobServiceClient: BlobServiceClient = new BlobServiceClient(
+      this._storageAccountUrl,
+      deviceCodeCredential
+    );
+
+    const startsOn: Date = new Date();
+    const expires: Date = new Date(Date.now() + SAS_TTL);
+    const key: ServiceGetUserDelegationKeyResponse = await blobServiceClient.getUserDelegationKey(
+      startsOn,
+      expires
+    );
+
+    const blobSasPermissions: BlobSASPermissions = new BlobSASPermissions();
+    blobSasPermissions.read = true;
+    blobSasPermissions.create = this._isCacheWriteAllowed;
+
+    const userDelegationKey: UserDelegationKey = key;
+    const queryParameters: SASQueryParameters = generateBlobSASQueryParameters(
+      {
+        startsOn: startsOn,
+        expiresOn: expires,
+        permissions: blobSasPermissions,
+        containerName: this._storageContainerName,
+        blobName: 'dummy-blob-name'
+      },
+      userDelegationKey,
+      this._storageAccountName
+    );
+
+    return queryParameters;
+  }
+
+  private _printMessageInBox(message: string, terminal: Terminal): void {
+    const boxWidth: number = Math.floor(Utilities.getConsoleWidth() / 2);
+    const maxLineLength: number = boxWidth - 10;
+
+    const wrappedMessage: string = Utilities.wrapWords(message, maxLineLength);
+    const wrappedMessageLines: string[] = wrappedMessage.split('\n');
+
+    // ╔═══════════╗
+    // ║  Message  ║
+    // ╚═══════════╝
+    terminal.writeLine(' ╔' + new Array(boxWidth - 3).join('═') + '╗ ');
+    for (const line of wrappedMessageLines) {
+      const trimmedLine: string = line.trim();
+      const padding: number = boxWidth - trimmedLine.length - 4;
+      const leftPadding: number = Math.floor(padding / 2);
+      const rightPadding: number = padding - leftPadding;
+      terminal.writeLine(
+        ' ║' +
+          new Array(leftPadding + 1).join(' ') +
```

Comment:
🤔 Isn't there a better way to do this?   (And isn't `new Array()` deprecated?)

Reply:
`' '.repeat(leftPadding + 1)` for example

Reply:
Looks like there's a `string.prototype.repeat` function.